Generics.md

```diff
@@ -1,164 +1,386 @@
+
+@override(texture_bind_group: 5, texture_type: f32)
+import util/sample_texture/{sample as sample_f32};
+```
+
+Importing the whole module is possible with the star (`*`) import.
+
+> NOTE: the declaration you provide as an override cannot access any of the declarations in the imported module. But it can access all declarations in the module where it is defined (see [examples](#examples)).
+
+### Overriding declarations via the linker
+
+This should be implementation-defined, but basically
+```ts
+const linker = new Linker();
+const linkerOptions = {
+  overrides: {
+    "some/file.wesl": { "texture_type": "u32" },
+    "some_library/some/file.wesl": { "group_offset": 10 },
   }
-  ```
+};
+let wgsl_source = linker.link("main.wesl", linkerOptions)
+```
 
-Here’s a more complicated case. reduce is parameterized by an element type (e.g. u32) and a binary operation, e.g. max()
+## (optional) Module-as-Type
 
-* ```
-    ./util.wgsl:
+We can allow a module to be passed as a type constraint. Then, a declaration with a module generic can be specialized with any other module that overrides all uninitialized declarations. This can help implement object-oriented or interface semantics.
 
-    export<E, BinOp> 
-    fn reduce(elems: array<E, 2>) -> E { 
-      return BinOp(elems[0], elems[1]); 
-    }
+Formally, a module `M1` is a subtype of module `M2` if,
+* for all declarations `D1` in `M1` that override a declaration `D2` in `M2`, `D1` is a subtype of `D2`.
+* all uninitialized declarations in `M2` are overridden in `M1`.
 
-* ```
-    ./main.wgsl:
+For this, a declaration can be decorated with the `@override()` attribute that specifies which declaration of which module it overrides.
+```rs
+// file animal.wesl
+fn say_hi();
+fn walk();
 
-    import ./ops/binOpMax<f32> as binOpMax;
-    import ./util/reduce<f32, binOpMax> as maxF32; 
+// file cat.wesl
+import animal/*;
+@override(animal::say_hi)
+fn meow() { ... }
+@override(animal::walk)
+fn walk() { ... }
+```
 
-    fn main() {
-      maxF32(a1);
-    }
-    ```
+```rs
+import animal as animal_mod;
+import cat;
 
-Note that you can import a generic function w/o providing parameters:
+@generic(A: animal_mod)
+fn simulate_animal() {
+  A::say_hi();
+  A::walk();
+  ...
+}
 
-* ```
-    ./util.wgsl:
+fn main() {
+  simulate_animal<cat>(); // calls cat::meow() and cat::walk()
+}
+```
 
-    import binOpMax from ./ops;   // no generic variable specified yet
+## Implementation
 
-    export fn reduceMax<E>(elems: array<E, 2>) -> E { 
-      return binOpMax<E>(elems[0], elems[1]); // generic value applied at call site
-    }
-    ```
+Type checking does not have to be enforced by the linker, especially for runtime linking where we care about performance.
+All of this proposal (except module-as-types) can be implemented with simple substitution.
+A compile-time linker or a IDE linter/langserver should implement type checking. Basically what Typescript does. 
```

Comment:
I think I know where you're going, but it's a little confusing because of course the TypeScript compiler does do type checking. (I think you mean that typical web development these days combines a fast transpiler that does no typechecking with a language server that does type checking)